### PR TITLE
Introduce SDL_HINT_EMSCRIPTEN_ASYNCIFY

### PR DESCRIPTION
**Pull request by Beuc**
See https://github.com/emscripten-core/emscripten/issues/10746

### CODE REVIEW

**Comment by Beuc** [comment]
Testing can be done through `emcc --clear-ports` and:
```
--- tools/ports/sdl2.py	2020-03-21 12:32:11.940361701 +0100
+++ tools/ports/sdl2.py	2020-04-10 23:26:37.196502209 +0200
@@ -6,7 +6,7 @@
 import os
 
 TAG = 'version_19'
-HASH = '5ee4c38ed8372ef1c9bca0d3826279be97bdb6150570b0e2b7c0e3a8c33f446b3ebd5df6385fccf6c25ac0fbb4ab00ed4728b52320951e648e38867cfd154795'
+HASH = '97082dba254ec6742354d18a997a75b76815172cfa16bd1c6bdfc081d89f71de48b7cf9b6af1e55b7cbbe9442f960c489bd54959d45942de5821d5608a3868a5'
 SUBDIR = 'SDL2-' + TAG
 
 
@@ -15,7 +15,7 @@
     return []
 
   # get the port
-  ports.fetch_project('sdl2', 'https://github.com/emscripten-ports/SDL2/archive/' + TAG + '.zip', SUBDIR, sha512hash=HASH)
+  ports.fetch_project('sdl2', 'https://github.com/Beuc/SDL2/archive/' + TAG + '.zip', SUBDIR, sha512hash=HASH)
   libname = ports.get_lib_name('libSDL2' + ('-mt' if settings.USE_PTHREADS else ''))
 
   def create():
```

**Comment by sy2002** [comment]
Thank you. Works for me.
Details: https://github.com/emscripten-core/emscripten/issues/10746#issuecomment-614014925

File: src/timer/unix/SDL_systimer.c [hunk 1]
```diff
@@ -192,8 +193,8 @@ void
 SDL_Delay(Uint32 ms)
 {
 #ifdef __EMSCRIPTEN__
-    if (emscripten_has_asyncify()) {
-        /* pseudo-synchronous pause */
+    if (emscripten_has_asyncify() && SDL_GetHintBoolean(SDL_HINT_EMSCRIPTEN_ASYNCIFY, SDL_TRUE)) {
```
Comment (sbc100) [inline src/timer/unix/SDL_systimer.c:196]: Since `emscripten_has_asyncify` is a JS call mabye put `SDL_GetHintBoolean` first to avoid doing it?

Comment (Beuc) [reply src/timer/unix/SDL_systimer.c:196]: I don't know. `emscripten_has_asyncify` is a static return, while SDL_GetHint/SDL_GetHintBoolean is a non-trivial function with a cost:
https://github.com/emscripten-ports/SDL2/blob/master/src/SDL_hints.c#L103

**Comment by Beuc** [comment]
Hi. What is the status? :)

**Comment by sbc100** [comment]
I'm not sure I'm the best reviewer, but rubberstamping since its seems to be what was agreed in issue

**Comment by sbc100** [comment]
I think you will need to find a contributor with write access to land this..

**Comment by kripken** [comment]
Landing with those two lgtms, thanks everyone!

**Comment by Daft-Freak** [comment]
Ah, meant to come back and merge this but forgot. Thanks @kripken!

**Comment by rcgordon** [comment]
Hey, I'm popping into a merged pull request to ask about this, sorry.

We are just about to fix this exact thing in upstream SDL: https://bugzilla.libsdl.org/show_bug.cgi?id=4997

Is it safe to have an SDL hint for this? I was under the assumption that Asyncify was something baked into the final build, not something you toggled at runtime, which would make an SDL hint dangerous here (and the call to emscripten_has_asyncify() by itself the correct approach). Am I misunderstanding how this works?

**Comment by Daft-Freak** [comment]
I think the hint is to avoid breaking the case where you're using Asyncify for something else, since the "automatic" support requires some extra setup.

(I guess this is one of the patches I forgot to push upstream... Should just be 4da3844f700a3f75f8e50e78ceee3c8062182019 and 5b09b696eb157aac701127c30b870b5e18082a92)

**Comment by Beuc** [comment]
Hi Ryan.

Please do not merge another implementation of asyncify support and hinting in the upstream repo.

This PR went through a few rounds of testing (see #70 and https://github.com/emscripten-core/emscripten/issues/10746), and was documented:
https://wiki.libsdl.org/SDL_HINT_EMSCRIPTEN_ASYNCIFY

Note: it would be nice to clarify where to contribute to SDL2 for Emscripten, from experience bugzilla gets stalled (e.g. 4203), while contributing here gets merged back from time to time, hence why I pushed code here.
Also upstream SDL is usually broken at a given point of time due to Emscripte/browser changes, so this port here is often the only one that works for the current web stack.
Maybe direct bugzilla entries about Emscripten here? :)

**Comment by rcgordon** [comment]
No worries, thanks for following up so quickly!

I've push these upstream now:

https://hg.libsdl.org/SDL/rev/e06e70f12286

https://hg.libsdl.org/SDL/rev/dd5a816a4d61

Thanks again!

**Comment by rcgordon** [comment]
> Please do not merge another implementation of asyncify support and hinting in the upstream repo.

Yeah, I came here because I thought the patch on Bugzilla seemed imcomplete, and Google brought me to this pull request.

> Note: it would be nice to clarify where to contribute to SDL2 for Emscripten,

I would personally prefer if the upstream was the primary target, to avoid forks, but things _do_ languish in Bugzilla from time to time, as our time and understanding of a problem are sometimes lacking (there's been several times where we didn't have any expertise on a specific issue and uncritically accepted someone's patch with disastrous consequences, for various platforms).

We gave @Daft-Freak commit access upstream, because he does great work and clearly understands the nuances of Emscripten better than we can hope to, and we trust anything he decides to push.

As to where bugs and patches should go? I'm happy with whatever you and Charlie and Kripken think is best. As long as they make it upstream eventually, work wherever you prefer: Github, bugzilla, or just yell at me on Twitter, it's all okay.

(I do occasionally work on the Emscripten port myself, and should probably run significant changes through here for commentary if nothing else.)

**Pull request merged**